File: couchbase-persistence/hello-service-java/hello-impl/src/main/java/org/example/hello/impl/HelloEventProcessor.java
package org.example.hello.impl;

import akka.Done;
import akka.stream.alpakka.couchbase.javadsl.CouchbaseSession;
import com.couchbase.client.java.document.JsonDocument;
import com.couchbase.client.java.document.json.JsonObject;
import com.lightbend.lagom.javadsl.persistence.AggregateEventTag;
import com.lightbend.lagom.javadsl.persistence.ReadSideProcessor;
import com.lightbend.lagom.javadsl.persistence.couchbase.CouchbaseReadSide;
import org.pcollections.PSequence;

import javax.inject.Inject;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

//#couchbase-begin
public class HelloEventProcessor extends ReadSideProcessor<HelloEvent> {

  static String DOC_ID = "users-actual-greetings";

  CouchbaseReadSide couchbaseReadSide;

  @Inject
  public HelloEventProcessor(CouchbaseReadSide couchbaseReadSide) {
    this.couchbaseReadSide = couchbaseReadSide;
  }

  @Override
  public ReadSideHandler<HelloEvent> buildHandler() {
    return couchbaseReadSide.<HelloEvent>builder("all-greetings")
        .setGlobalPrepare(this::globalPrepare)
        .setEventHandler(HelloEvent.GreetingMessageChanged.class, this::processGreetingMessageChanged)
        .build();
  }

  private CompletionStage<Done> globalPrepare(CouchbaseSession session) {
    return
        session.get(DOC_ID).thenComposeAsync(doc -> {
          if (doc.isPresent()) {
            return CompletableFuture.completedFuture(Done.getInstance());
          }
          return session.insert(JsonDocument.create(DOC_ID, JsonObject.empty()))
              .thenApply(ignore -> Done.getInstance());
        });
  }

  private CompletionStage<Done> processGreetingMessageChanged(CouchbaseSession session, HelloEvent.GreetingMessageChanged evt) {
    return session.get(DOC_ID).thenComposeAsync(docOpt -> {
      JsonDocument doc = docOpt.get(); // document should've been created in globalPrepare

      doc.content().put(evt.name, evt.message);

      return session.upsert(doc);
    }).thenApply(ignore -> Done.getInstance());
  }

  @Override
  public PSequence<AggregateEventTag<HelloEvent>> aggregateTags() {
    return HelloEvent.TAG.allTags();
  }
}
//#couchbase-end

File: couchbase-persistence/couchbase-persistence-scala-sbt/hello-impl/src/main/scala/org/example/hello/impl/HelloEventProcessor.scala
package org.example.hello.impl

import akka.Done
import akka.actor.ActorSystem
import akka.stream.alpakka.couchbase.scaladsl.CouchbaseSession
import com.couchbase.client.java.document.json.JsonObject
import com.couchbase.client.java.document.JsonDocument
import com.lightbend.lagom.scaladsl.persistence.{
  AggregateEventTag,
  EventStreamElement,
  ReadSideProcessor
}
import com.lightbend.lagom.scaladsl.persistence.couchbase.CouchbaseReadSide

import scala.concurrent.{ExecutionContext, Future}

//#couchbase-begin
object HelloEventProcessor {
  private[impl] val UserGreetingsDocId = "users-actual-greetings"
}

class HelloEventProcessor(
  couchbaseReadSide: CouchbaseReadSide,
  actorSystem: ActorSystem
)(implicit executionContext: ExecutionContext)
    extends ReadSideProcessor[HelloEvent] {

  import HelloEventProcessor.UserGreetingsDocId

  override def buildHandler(): ReadSideProcessor.ReadSideHandler[HelloEvent] = {

    couchbaseReadSide
      .builder[HelloEvent]("all-greetings")
      .setGlobalPrepare(createDocument _)
      .setEventHandler[GreetingMessageChanged](processGreetingMessageChanged _)
      .build()
  }
  private def createDocument(session: CouchbaseSession): Future[Done] =
    session.get(UserGreetingsDocId).flatMap {
      case Some(doc) => Future.successful(Done)
      case None =>
        session
          .upsert(JsonDocument.create(UserGreetingsDocId, JsonObject.empty()))
          .map(_ => Done)
    }

  private def processGreetingMessageChanged(
    session: CouchbaseSession,
    ese: EventStreamElement[GreetingMessageChanged]
  ): Future[Done] =
    session
      .get(UserGreetingsDocId)
      .flatMap { maybeDoc =>
        val json = maybeDoc match {
          case Some(doc) => doc.content()
          case None      => JsonObject.create();
        }
        val evt = ese.event
        json.put(evt.name, evt.message)
        session.upsert(JsonDocument.create(UserGreetingsDocId, json))
      }
      .map(_ => Done)

  override def aggregateTags: Set[AggregateEventTag[HelloEvent]] =
    HelloEvent.Tag.allTags
}
//#couchbase-end
